Update parsing md & mention it in plugin tuto mlg files (#20669)
rocq-prover/rocq: dev/doc/parsing.md

in one file and the parser action routines and other OCaml code are omitted.

`*.mlg` files contain the following types of nodes (See `node` in the yacc grammar).  This part is
very specific to Rocq (not so similar to Camlp5).

Note that you can reverse engineer many details by comparing the `.mlg` input file with
the `.ml` generated by `coqpp`.

### OCaml code

OCaml code enclosed in curly braces, which is copied verbatim to the generated `*.ml` file

### Comments

> Comments in the `*.mlg` file in the form `(* … *)`, which are not copied
> to the generated `*.ml` file.  Comments in OCaml code are preserved.

Reviewer: ```suggestion
Comments in the `*.mlg` file are in the form `(* … *)`. They are not copied
to the generated `*.ml` file.  Comments in OCaml code are preserved.
```

END
```

Global entries should be available in the current OCaml scope,
eg `open Procq.Prim` makes `qualid` available. (qualified entry names are not supported)

`GRAMMAR EXTEND` should be used only for large syntax additions.  To add new commands
and tactics, use these instead:

- `VERNAC COMMAND EXTEND` (and variant `VERNAC { entry } EXTEND`) to add new commands
- `TACTIC EXTEND` to add new ltac1 tactics
- `ARGUMENT EXTEND` to add new arguments for `TACTIC EXTEND` and `VERNAC EXTEND`
- `VERNAC ARGUMENT EXTEND` to add new arguments for `VERNAC EXTEND`

These constructs provide essential semantic information that's provided in a more complex,
less readable way with `GRAMMAR EXTEND`.

### VERNAC EXTEND

`VERNAC COMMAND EXTEND` - adds new command syntax by adding productions to the
`command` entry.

`VERNAC { entry } EXTEND` - adds new command syntax to the `entry` entry.
This is typically used for proof modes (see also `Pvernac.register_proof_mode` and the user doc for proof modes).

Simple example:

```
VERNAC COMMAND EXTEND CmdName CLASSIFIED AS SIDEFF
| [ "Cmd" arg(a) ] -> { do_interp a }
END
```
defines a command composed of the string `Cmd` followed by argument `wit_arg`.

> Running the command runs `do_interp a` where `a` is the result of parsing `arg`.

Reviewer: ```suggestion
Running the command calls `do_interp a` where `a` is the result of parsing `arg`.
```

Example with all optional elements:
```
VERNAC COMMAND EXTEND CmdName CLASSIFIED AS SIDEFF
| #[ attr ] ![ state ] [ "Cmd" arg(a) ]
  => { rule_classif a }
  SYNTERP AS synv { do_synterp attr a }
  -> { do_interp attr a synv }
END
```

Productions here are represented with [alternate syntax](#alternate-production-syntax), described later.

Nonterminals in `VERNAC EXTEND` parsing rules (e.g. `arg` in the example)
are arguments, possibly with [nonterminal modifiers](#nonterminal-modifiers).
The name prefixed by `wit_` must be available in the current OCaml scope.

New commands should be added using this construct rather than `GRAMMAR EXTEND` so
they are correctly registered, such as having the correct command classifier.

#### Extend name

Each `VERNAC EXTEND` must have a unique name ("CmdName" in the example)
(unique for the plugin, you will get an anomaly at runtime if it's
not).

#### Classification

> Commands must be "classified" for the STM. This means generating a

Reviewer: STM?

Author: "State Transaction Machine", it drives command execution in rocq compile, rocq top and coqidetop (not vsrocq and coq-lsp).
It is problematic so we would like to remove it someday but have not managed so far.

> Commands must be "classified" for the STM. This means generating a
> `Vernacextend.vernac_classification` (see comments on its
> declaration for the meaning of the different values), which can be
> done with

Reviewer: ```suggestion
Commands must be "classified" for the STM.
Classifying a command is generating a
`Vernacextend.vernac_classification` (see comments on its
declaration for the meaning of the different values).
This is done by appending one of the following annotations to the `VERNAC COMMAND EXTEND CmdName` grammar extension:
```

the parsing rule. `x` and `y` will be bound to the parsed attribute
in the interpretation, `a` and `b` are qualified identifiers of type
`'att Attributes.attribute`. Punning `#[ x ]` for `#[ x = x ]` is supported.
You can delay attribute interpretation (e.g. if it depends on the command arguments)
by using `Attributes.raw_attributes`.

#### Synterp

If the command modifies the parser or modifies state which may be
used to change the parser in a later command, these operations must
be separated from the main interpretation and done in the "synterp"
phase. The "synterp" phase does not have access to the "interp"
state (which notably includes the global env).

This is done with `SYNTERP AS synv { synterp }`: `synterp` will be
evaluated in the synterp phase with the command arguments and attributes bound,
and its result will be bound to `synv` for the interp phase.

### TACTIC EXTEND

`TACTIC EXTEND` - adds new tactic (Ltac1) syntax by adding productions to `simple_tactic`.
For example:

```
TACTIC EXTEND field_lookup
| [ "field_lookup" tactic(f) "[" constr_list(lH) "]" ne_constr_list(lt) ] ->
      { let (t,l) = List.sep_last lt in field_lookup f lH l t }
END
```

> adds a new nonterminal `btauto`.

Reviewer: Isn't it `field_lookup`?
